### PR TITLE
MatchErr is set to ErrNotFound if NotFoundHandler is used

### PR DESCRIPTION
**Pull request by roobre**
As discussed in #293, it should be useful if `RouteMatch.MatchErr` were set to an error if no match is found, even if some subrouters have a custom `NotFoundHandler`.

### CODE REVIEW

File: mux_test.go [hunk 1]
```diff
@@ -1982,3 +1982,54 @@ func TestNoMatchMethodErrorHandler(t *testing.T) {
...
+	s := r.PathPrefix("/sub/").Subrouter()
+	s.HandleFunc("/", emptyHandler)
+
+	// Regular 404 not found
+	req, _ := http.NewRequest("GET", "/sub/whatever", nil)
+	match := new(RouteMatch)
+	matched := r.Match(req, match)
+
+	if matched {
+		t.Error("Subrouter should not have matched that")
```
Comment (elithrar) [inline mux_test.go]: I would add some context - `”subrouter should not have matched: got %v”, route”` to aid debugging.

File: mux_test.go [hunk 2]
```diff
@@ -1982,3 +1982,54 @@ func TestNoMatchMethodErrorHandler(t *testing.T) {
...
+	// Now lets add a 404 handler to subrouter
+	s.NotFoundHandler = http.NotFoundHandler()
+	req, _ = http.NewRequest("GET", "/sub/whatever", nil)
+
+	// Test the subrouter first
+	match = new(RouteMatch)
+	matched = s.Match(req, match)
+	// Now we should get a match
+	if !matched {
+		t.Error("Subrouter should have matched that")
```
Comment (elithrar) [inline mux_test.go]: Here too.

File: mux_test.go [hunk 3]
```diff
@@ -1982,3 +1982,54 @@ func TestNoMatchMethodErrorHandler(t *testing.T) {
...
+		t.Errorf("Expected ErrNotFound MatchErr, but was %v", match.MatchErr)
+	}
+
+	// Now test the parent (MatchErr should propagate)
+	match = new(RouteMatch)
+	matched = r.Match(req, match)
+
+	// Now we should get a match
+	if !matched {
+		t.Error("Subrouter should have matched that")
```
Comment (elithrar) [inline mux_test.go]: Here too.

**Comment by roobre** [comment]
I've added the matching route where none is expected but one is got. I've also moved the last two test cases above the private functions, which I think is more appropriate.

The go fmt issue with travis is bugging me a bit, but afaik it's not my fault. It appeared when I merged upstream :/

**Comment by elithrar** [comment]
Thanks for this. Let me see why gofmt is failing [only] on tip.

**Comment by elithrar** [comment]
Rebase against master to pull in https://github.com/gorilla/mux/pull/312

File: route.go [hunk 4]
```diff
@@ -72,7 +72,11 @@ func (r *Route) Match(req *http.Request, match *RouteMatch) bool {
 		return false
 	}
 
-	match.MatchErr = nil
+	if match.MatchErr == ErrMethodMismatch {
+		// We found a non-mismatching route, clear MatchErr
```
Comment (kisielk) [inline route.go]: non-mismatching is a bit of a double negative :) how about just "matching" so it's less confusing?

**Comment by kisielk** [comment]
LGTM other than comment

File: route.go (hunk 4 above)
Comment (roobre) [reply route.go]: Hm, I used that wording to make clear that the ErrMethod*Mismatch* found on the previous iteration no longer applies. But I guess I can reword it a bit.

I'll do it tomorrow tho, its 2AM here and I'm quite asleep.

**Comment by kisielk** [comment]
Thanks! I don't mean to be too nitpicky, but the code base is already pretty hard to follow so any little bit helps.

**Comment by roobre** [comment]
@kisielk No prob! I'm not a native speaker, so I can make some grammar mistakes here and there. Also, I'd do the same on projects I mantain :P
